## 1 Descarga


La carga de información se realizó conn con el cuaderno 02downloadDataV2.0.ipynb


## 1 Preprocessado de Datos

In [27]:
%pip install pandas
%pip install scikit-learn
%pip install tqdm

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [40]:
import pandas as pd
import os
from tqdm import tqdm
from sklearn.model_selection import train_test_split



In [ ]:
#carga de informacion
df = pd.read_csv('dataset.csv', sep=';')



Listado de las columnas presentes en el archivo `filtered_data.csv` 

1. **id**: Identificador único de la propiedad.
2. **listing_url**: URL del anuncio de la propiedad en AirBnB.
3. **scrape_id**: Identificador único para la tarea de scraping.
4. **last_scraped**: Fecha en la que se realizó el último scraping de los datos.
5. **source**: Fuente de los datos (por ejemplo, "city scrape").
6. **picture_url**: URL de la imagen de la propiedad.
7. **host_id**: Identificador único del anfitrión.
8. **host_url**: URL del perfil del anfitrión.
9. **host_name**: Nombre del anfitrión.
10. **host_since**: Fecha en la que el anfitrión se unió a AirBnB.
11. **host_location**: Ubicación del anfitrión (por ejemplo, ciudad o región).
12. **host_about**: Descripción del anfitrión.
13. **host_response_time**: Tiempo de respuesta del anfitrión a los mensajes.
14. **host_response_rate**: Tasa de respuesta del anfitrión.
15. **host_acceptance_rate**: Tasa de aceptación de las solicitudes del anfitrión.
16. **host_is_superhost**: Indica si el anfitrión es un "superhost".
17. **host_thumbnail_url**: URL de la imagen pequeña del perfil del anfitrión.
18. **host_picture_url**: URL de la imagen completa del perfil del anfitrión.
19. **host_neighbourhood**: Barrio o vecindad del anfitrión.
20. **host_listings_count**: Número de propiedades que el anfitrión tiene listadas.
21. **host_total_listings_count**: Total de propiedades que el anfitrión ha listado a lo largo del tiempo.
22. **host_verifications**: Verificaciones del anfitrión (por ejemplo, email, teléfono).
23. **host_has_profile_pic**: Indica si el anfitrión tiene foto de perfil.
24. **host_identity_verified**: Indica si la identidad del anfitrión ha sido verificada.
25. **neighbourhood**: Barrio o vecindad de la propiedad.
26. **neighbourhood_cleansed**: Barrio de la propiedad después de limpiarlo de inconsistencias.
27. **neighbourhood_group_cleansed**: Grupo de barrios de la propiedad después de limpiarlo de inconsistencias.
28. **latitude**: Latitud de la ubicación de la propiedad.
29. **longitude**: Longitud de la ubicación de la propiedad.
30. **property_type**: Tipo de propiedad (por ejemplo, "Apartamento", "Casa").
31. **room_type**: Tipo de habitación (por ejemplo, "Habitación privada", "Habitación compartida").
32. **accommodates**: Número máximo de personas que puede alojar la propiedad.
33. **bathrooms**: Número de baños de la propiedad.
34. **bathrooms_text**: Descripción textual de los baños de la propiedad.
35. **bedrooms**: Número de habitaciones de la propiedad.
36. **beds**: Número de camas de la propiedad.
37. **amenities**: Lista de servicios y comodidades que ofrece la propiedad.
38. **price**: Precio de la propiedad por noche.
39. **minimum_nights**: Número mínimo de noches para reservar la propiedad.
40. **maximum_nights**: Número máximo de noches para reservar la propiedad.
41. **minimum_minimum_nights**: Valor mínimo de noches mínimas en las propiedades.
42. **maximum_minimum_nights**: Valor máximo de noches mínimas en las propiedades.
43. **minimum_maximum_nights**: Valor mínimo de noches máximas en las propiedades.
44. **maximum_maximum_nights**: Valor máximo de noches máximas en las propiedades.
45. **minimum_nights_avg_ntm**: Promedio de noches mínimas por mes.
46. **maximum_nights_avg_ntm**: Promedio de noches máximas por mes.
47. **calendar_updated**: Fecha de la última actualización del calendario.
48. **has_availability**: Indica si la propiedad está disponible para alquilar.
49. **availability_30**: Disponibilidad de la propiedad para los próximos 30 días.
50. **availability_60**: Disponibilidad de la propiedad para los próximos 60 días.
51. **availability_90**: Disponibilidad de la propiedad para los próximos 90 días.
52. **availability_365**: Disponibilidad de la propiedad para todo el año.
53. **calendar_last_scraped**: Fecha en la que se raspó por última vez el calendario.
54. **number_of_reviews**: Número total de reseñas de la propiedad.
55. **number_of_reviews_ltm**: Número de reseñas en los últimos 12 meses.
56. **number_of_reviews_l30d**: Número de reseñas en los últimos 30 días.
57. **first_review**: Fecha de la primera reseña.
58. **last_review**: Fecha de la última reseña.
59. **review_scores_rating**: Puntuación general de la propiedad en las reseñas.
60. **review_scores_accuracy**: Puntuación de exactitud de las reseñas.
61. **review_scores_cleanliness**: Puntuación de limpieza de las reseñas.
62. **review_scores_checkin**: Puntuación de la experiencia de check-in.
63. **review_scores_communication**: Puntuación de la comunicación con el anfitrión.
64. **review_scores_location**: Puntuación de la ubicación de la propiedad.
65. **review_scores_value**: Puntuación del valor de la propiedad.
66. **license**: Licencia de la propiedad, si está disponible.
67. **instant_bookable**: Indica si la propiedad es reservable instantáneamente.
68. **calculated_host_listings_count**: Número total de propiedades del anfitrión.
69. **calculated_host_listings_count_entire_homes**: Número de propiedades del anfitrión que son casas completas.
70. **calculated_host_listings_count_private_rooms**: Número de propiedades del anfitrión que son habitaciones privadas.
71. **calculated_host_listings_count_shared_rooms**: Número de propiedades del anfitrión que son habitaciones compartidas.
72. **reviews_per_month**: Promedio de reseñas por mes de la propiedad.

---

Estas columnas contienen información clave sobre las propiedades de AirBnB y sus anfitriones, desde los detalles de la propiedad hasta las puntuaciones de las reseñas. Si necesitas más información sobre alguna columna o cómo trabajar con estos datos, no dudes en preguntar.

In [37]:
df.dtypes

id                                                int64
listing_url                                      object
scrape_id                                         int64
last_scraped                                     object
source                                           object
picture_url                                      object
host_id                                           int64
host_url                                         object
host_name                                        object
host_since                                       object
host_location                                    object
host_about                                       object
host_response_time                               object
host_response_rate                               object
host_acceptance_rate                             object
host_is_superhost                                object
host_thumbnail_url                               object
host_picture_url                                

In [26]:
# Contar la cantidad de duplicados en la columna 'id'
cantidad_duplicados = df['id'].duplicated().sum()
print(f'Cantidad de duplicados en la columna "id": {cantidad_duplicados}')


directorio = 'imagenes'  # Cambia esto a la ruta correcta de tu directorio
archivos_jpg = [archivo for archivo in os.listdir(directorio) if archivo.lower().endswith('.jpg')]
cantidad_jpg = len(archivos_jpg)
print(f'Cantidad de archivos .jpg en el directorio "{directorio}": {cantidad_jpg}')



Cantidad de duplicados en la columna "id": 3599
Cantidad de archivos .jpg en el directorio "imagenes": 3109


NOTA: se ahn desacrgado 3109 imagees , se separaran lso registros que tiene imagenes para proceder con la practica

In [41]:
import os
import pandas as pd
from tqdm import tqdm

directorio_imagenes = 'imagenes'  # Cambia esto a la ruta correcta de tu directorio de imágenes
filas_descargadas = []
ids_procesados = set()

for index, row in tqdm(df.iterrows(), total=df.shape[0], desc="Procesando imágenes"):
    archivo_imagen = str(row['id']) + '.jpg'    
    if os.path.isfile(os.path.join(directorio_imagenes, archivo_imagen)):
        # Verificar si el ID ya ha sido procesado
        if row['id'] not in ids_procesados:
            filas_descargadas.append(row)
            ids_procesados.add(row['id'])

df_descargados = pd.DataFrame(filas_descargadas)

# Mostrar el DataFrame de imágenes descargadas
print(f"Registros separadaos {len(df_descargados)} .")
display(df_descargados.head(4))  # Muestra las primeras filas de df_descargados

df_descargados.to_csv('03_data_descargada.csv', sep='|', index=False)
df_descargados.to_pickle('03_data_descargada.pkl')


Procesando imágenes: 100%|██████████| 7199/7199 [00:00<00:00, 34002.61it/s]

Registros separadaos 3109 .


id                        listing_url       scrape_id last_scraped  \
0  3781  https://www.airbnb.com/rooms/3781  20231218233145   2023-12-19   
1  5506  https://www.airbnb.com/rooms/5506  20231218233145   2023-12-19   
2  6695  https://www.airbnb.com/rooms/6695  20231218233145   2023-12-19   
3  8789  https://www.airbnb.com/rooms/8789  20231218233145   2023-12-19   

        source                                        picture_url  host_id  \
0  city scrape  https://a0.muscache.com/pictures/24670/b2de044...     4804   
1  city scrape  https://a0.muscache.com/pictures/miso/Hosting-...     8229   
2  city scrape  https://a0.muscache.com/pictures/38ac4797-e7a4...     8229   
3  city scrape  https://a0.muscache.com/pictures/miso/Hosting-...    26988   

                                  host_url host_name  host_since  \
0   https://www.airbnb.com/users/show/4804     Frank  2008-12-03   
1   https://www.airbnb.com/users/show/8229     Terry  2009-02-19   
2   https://www.airbnb.com/users/show/8229     Terry  2009-02-19   
3  https://www.airbnb.com/users/show/26988      Anne  2009-07-22   

                  host_location  \
0  Massachusetts, United States   
1                    Boston, MA   
2                    Boston, MA   
3                    Boston, MA   

                                          host_about host_response_time  \
0  My wife and I and grown children frequently oc...       within a day   
1              Relaxed,  Easy going, Accommodating.      within an hour   
2              Relaxed,  Easy going, Accommodating.      within an hour   
3  I manage a number of furnished units around Bo...       within a day   

  host_response_rate host_acceptance_rate host_is_superhost  \
0                90%                  29%                 t   
1               100%                 100%                 t   
2               100%                 100%                 t   
3                90%                  50%                 t   

                                  host_thumbnail_url  \
0  https://a0.muscache.com/im/users/4804/profile_...   
1  https://a0.muscache.com/im/users/8229/profile_...   
2  https://a0.muscache.com/im/users/8229/profile_...   
3  https://a0.muscache.com/im/pictures/user/d7d71...   

                                    host_picture_url host_neighbourhood  \
0  https://a0.muscache.com/im/users/4804/profile_...        East Boston   
1  https://a0.muscache.com/im/users/8229/profile_...            Roxbury   
2  https://a0.muscache.com/im/users/8229/profile_...            Roxbury   
3  https://a0.muscache.com/im/pictures/user/d7d71...        Beacon Hill   

   host_listings_count  host_total_listings_count  host_verifications  \
0                    4                          5  ['email', 'phone']   
1                   11                         14  ['email', 'phone']   
2                   11                         14  ['email', 'phone']   
3                   13                         31  ['email', 'phone']   

  host_has_profile_pic host_identity_verified  \
0                    t                      f   
1                    t                      t   
2                    t                      t   
3                    t                      t   

                          neighbourhood neighbourhood_cleansed  \
0  Boston, Massachusetts, United States            East Boston   
1  Boston, Massachusetts, United States                Roxbury   
2  Boston, Massachusetts, United States                Roxbury   
3  Boston, Massachusetts, United States            Beacon Hill   

   neighbourhood_group_cleansed  latitude  longitude       property_type  \
0                           NaN  42.36413  -71.02991  Entire rental unit   
1                           NaN  42.32844  -71.09581  Entire guest suite   
2                           NaN  42.32802  -71.09387        Entire condo   
3                           NaN  42.35867  -71.06307  Entire rental unit   

         room_type  accommodates  bathrooms bath

In [42]:
#Eliminamos las columnas que no aportarian al modelo
#o que tienen ideas semejantes como host_id(que se mantien) y se quietan host_url,host_picture_url,etc
#o columnas vacias
#o columans que no aportarian por ser unitarios como Licesse


# Definir las columnas a eliminar
columnas_a_eliminar = [
    'listing_url', 'scrape_id', 'last_scraped', 'source', 'picture_url',
    'host_url', 'host_name', 'host_thumbnail_url', 'host_picture_url',
    'neighbourhood', 'neighbourhood_cleansed', 'neighbourhood_group_cleansed',
    'bathrooms', 'bedrooms', 'amenities', 'calendar_updated', 'calendar_last_scraped', 'license'
]

df_descargados_sin_basura = df_descargados.drop(columns=columnas_a_eliminar, errors='ignore')

In [44]:
df_descargados_sin_basura.dtypes

id                                                int64
host_id                                           int64
host_since                                       object
host_location                                    object
host_about                                       object
host_response_time                               object
host_response_rate                               object
host_acceptance_rate                             object
host_is_superhost                                object
host_neighbourhood                               object
host_listings_count                               int64
host_total_listings_count                         int64
host_verifications                               object
host_has_profile_pic                             object
host_identity_verified                           object
latitude                                        float64
longitude                                       float64
property_type                                   

In [43]:


# Suponiendo que 'df' es tu DataFrame cargado
# Separar las características (X) y la variable objetivo (y)
X = df_descargados_sin_basura.drop(columns=['price'])  # Suponiendo que 'price' es la variable objetivo
y = df_descargados_sin_basura['price']

# Paso 1: Dividir en conjunto de entrenamiento (train) y conjunto de prueba (test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Paso 2: Dividir el conjunto de entrenamiento (train) en entrenamiento y validación (val)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

# Mostrar las dimensiones de cada conjunto
print(f"Training set: {X_train.shape[0]} samples")
print(f"Validation set: {X_val.shape[0]} samples")
print(f"Test set: {X_test.shape[0]} samples")

Training set: 1989 samples
Validation set: 498 samples
Test set: 622 samples


## 2 Modelado: Modelo 1D/tabular 

## 2 Modelado: Modelo 2D /Imagenes

## 3 Fusión de modelos: Estrategia late-fusion

## 3 Fusión de modelos: Estrategia early-fusion